include/rmm/mr/device/aligned_resource_adaptor.hpp

```diff
@@ -0,0 +1,175 @@
+ * using the given alignment size.
+ *
+ * @tparam Upstream Type of the upstream resource used for allocation/deallocation.
+ */
+template <typename Upstream>
+class aligned_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct an aligned resource adaptor using `upstream` to satisfy allocation requests.
+   *
+   * If the allocation size is smaller or equal to the specified alignment size, the default
+   * alignment from `upstream` is used; if the allocation size is larger, it is aligned up to the
+   * specified alignment size.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory.
+   */
+  explicit aligned_resource_adaptor(Upstream* upstream,
+                                    std::optional<std::size_t> allocation_alignment = std::nullopt)
+    : upstream_{upstream},
+      allocation_alignment_{allocation_alignment.value_or(default_allocation_alignment)}
+  {
+    RMM_EXPECTS(nullptr != upstream, "Unexpected null upstream resource pointer.");
+  }
+
+  aligned_resource_adaptor()                                = delete;
+  ~aligned_resource_adaptor() override                      = default;
+  aligned_resource_adaptor(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor(aligned_resource_adaptor&&)      = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor const&) = delete;
+  aligned_resource_adaptor& operator=(aligned_resource_adaptor&&) = delete;
+
+  /**
+   * @brief Get the upstream memory resource.
+   *
+   * @return Upstream* pointer to a memory resource object.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @copydoc rmm::mr::device_memory_resource::supports_streams()
+   */
+  [[nodiscard]] bool supports_streams() const noexcept override
+  {
+    return upstream_->supports_streams();
+  }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  [[nodiscard]] bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_streams();
+  }
+
+ private:
+  static constexpr std::size_t default_allocation_alignment = 4096;
+
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream resource with the specified
+   * alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    return upstream_->allocate(align(bytes), stream);
```

Comment:
I don't see how allocating an aligned value of bytes results in an aligned pointer. If the alignment is greater than the upstream allocator, I don't see a way around allocating `bytes + alignement_value` for  padding and adjusting the pointer to be aligned. Then you'll have to store a mapping of the returned pointer and the upstream pointer so you can free it properly.

Reply:
Done.